### PR TITLE
Refactor area, grid reference and x & y auto population

### PR DESCRIPTION
**Pull request by Cruikshanks**
https://eaflood.atlassian.net/browse/RUBY-483

It was spotted during testing of the area lookup integration that some of the extra functions that happens when we add an address had stopped working.

After some investigation it looks like the work we have been doing to refactor and simplify the forms used for addresses has broken a few things 😢

This change is doing 2 things

- getting them working again (obviously!)
- refactoring the way we do these things to better fit with the refactored forms

On that last point as we're now not trying to cram stuff into `BaseAddressFoo` forms it also doesn't make sense anymore to shove all these extra functions into `TransientAddress` as class methods.

Which is why this change covers moving them to where they are used, and calling them in `AfterCreate` hooks as that better fits what we're trying to do (create the address _then_ update some other properties based in the address data we have).

### CODE REVIEW

File: app/forms/concerns/waste_exemptions_engine/can_create_address_from_finder_data.rb [hunk 1]
```diff
@@ -11,7 +11,7 @@ def create_address(selected_address_uprn, type)
         data = temp_addresses.detect { |address| address["uprn"] == selected_address_uprn.to_i }
         return unless data
 
-        address = TransientAddress.create_from_address_finder_data(data, type)
+        address = TransientAddress.create_from_address_finder_data(data, TransientAddress.address_types[type])
```
Comment (cintamani) [inline app/forms/concerns/waste_exemptions_engine/can_create_address_from_finder_data.rb]: Just FYI: https://github.com/DEFRA/waste-exemptions-engine/pull/383/files#diff-04df5204897e7bf26c6002f14ff9e658 :D

File: app/models/waste_exemptions_engine/transient_address.rb [hunk 2]
```diff
@@ -53,71 +39,104 @@ def self.create_from_grid_reference_data(data, address_type)
...
+    # rubocop:disable Metrics/PerceivedComplexity
+    def update_site_details
+      return unless site?
+
+      update_x_and_y_from_grid_reference if x.blank? || y.blank?
+      update_x_and_y_from_postcode if x.blank? || y.blank?
+      update_grid_reference_from_x_and_y if grid_reference.blank?
+      update_area_from_x_and_y if area.blank?
+
+      save!
```
Comment (cintamani) [inline app/models/waste_exemptions_engine/transient_address.rb]: Can we create a service that deal with this stuff? Is a lot of code for this model to deal with.

**Comment by Cruikshanks** [comment]
No one seems to like my `update_site_details` method ☹️ . Is there anything meaningful I can do other than arbitrarily splitting them up?

File: spec/models/waste_exemptions_engine/transient_address_spec.rb [hunk 3]
```diff
@@ -23,6 +24,147 @@ module WasteExemptionsEngine
...
 
+    describe "hooks" do
+      context "creating a site address" do
+        context "populated from a grid reference" do
+          before(:context) { VCR.insert_cassette("site_address_area_grid_reference", allow_playback_repeats: true) }
+          after(:context) { VCR.eject_cassette }
+
+          subject(:transient_address) { create(:transient_address, :site_using_grid_reference) }
+          it "updates the x & y fields" do
+            subject.reload
```
Comment (cintamani) [inline spec/models/waste_exemptions_engine/transient_address_spec.rb]: Why do we need this reload? 🤔

File: app/models/waste_exemptions_engine/transient_address.rb [hunk 4]
```diff
@@ -13,6 +13,8 @@ class TransientAddress < ActiveRecord::Base
     enum address_type: { unknown: 0, operator: 1, contact: 2, site: 3 }
     enum mode: { unknown_mode: 0, lookup: 1, manual: 2, auto: 3 }
 
+    after_create :update_site_details
```
Comment (cintamani) [inline app/models/waste_exemptions_engine/transient_address.rb]: Shall we: 
```suggestion
    after_create :update_site_details, if: { site_address? }
```

File: app/forms/concerns/waste_exemptions_engine/can_create_address_from_finder_data.rb (hunk 1 above)
Comment (Cruikshanks) [reply app/forms/concerns/waste_exemptions_engine/can_create_address_from_finder_data.rb]: Well we'll just see who gets there first!

File: spec/models/waste_exemptions_engine/transient_address_spec.rb (hunk 3 above)
Comment (Cruikshanks) [reply spec/models/waste_exemptions_engine/transient_address_spec.rb]: Well lets find out!

File: app/forms/concerns/waste_exemptions_engine/can_create_address_from_finder_data.rb (hunk 1 above)
Comment (cintamani) [reply app/forms/concerns/waste_exemptions_engine/can_create_address_from_finder_data.rb]: That's unfair! It depends on you reviewing my PRs XD

File: app/models/waste_exemptions_engine/transient_address.rb (hunk 2 above)
Comment (Cruikshanks) [reply app/models/waste_exemptions_engine/transient_address.rb]: 😩 Yes boss!

poodr

**Comment by Cruikshanks** [comment]
Ready for round 2!

File: app/models/concerns/waste_exemptions_engine/can_check_easting_and_northing_values.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+      return false unless valid_coordinate?(northing)
+
+      true
+    end
+
+    def valid_coordinate?(coordinate)
+      return false unless coordinate.is_a?(Numeric)
+      return false unless coordinate.positive?
+
+      true
```
Comment (cintamani) [inline app/models/concerns/waste_exemptions_engine/can_check_easting_and_northing_values.rb]:
```suggestion
      coordinate.is_a?(Numeric) && coordinate.positive?
```

File: app/models/concerns/waste_exemptions_engine/can_check_easting_and_northing_values.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+module WasteExemptionsEngine
+  # Used to perform a basic check on easting and northing (x & y) values
+  module CanCheckEastingAndNorthingValues
+    extend ActiveSupport::Concern
+
+    private
+
+    def valid_coordinates?(easting, northing)
+      return false unless valid_coordinate?(easting)
+      return false unless valid_coordinate?(northing)
```
Comment (cintamani) [inline app/models/concerns/waste_exemptions_engine/can_check_easting_and_northing_values.rb]:
```suggestion
      valid_coordinate?(easting) && valid_coordinate?(northing)
```

File: app/models/waste_exemptions_engine/transient_address.rb [hunk 7]
```diff
@@ -53,71 +40,32 @@ def self.create_from_grid_reference_data(data, address_type)
...
-      # Add the grid reference for sites entered using an address
-      data = update_grid_reference_from_xy(data) unless mode == TransientAddress.modes[:auto]
+    private
 
-      # Add the EA administrative area
-      data = update_area_from_xy(data)
+    def update_site_details
+      update_x_and_y
+      update_grid_reference
+      update_area_from_x_and_y if area.blank?
```
Comment (cintamani) [inline app/models/waste_exemptions_engine/transient_address.rb]: Why do we check if `area.blank?` in-line here, and we check `valid_coordinates?(x, y)` in the `update_x_and_y` method instead? Either we do the check before, or in, or after (in the service). In any case, let's just be consistent :) 
Also the naming of the method, we are not `updating`, we are just `assigning`

File: app/services/waste_exemptions_engine/determine_area_service.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    def run(easting:, northing:)
+      return unless valid_coordinates?(easting, northing)
+
+      lookup_area(easting, northing)
+    end
+
+    private
+
+    def lookup_area(easting, northing)
+      response = DefraRuby::Area::PublicFaceAreaService.run(easting, @orthing)
```
Comment (cintamani) [inline app/services/waste_exemptions_engine/determine_area_service.rb]:
```suggestion
      response = DefraRuby::Area::PublicFaceAreaService.run(easting, northing)
```
☝️ are we missing test coverage here?

File: app/services/waste_exemptions_engine/determine_easting_and_northing_service.rb [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+module WasteExemptionsEngine
+  class DetermineEastingAndNorthingService < BaseService
+    def run(grid_reference:, postcode:)
+      @result = { easting: nil, northing: nil }
+      return @result unless valid_arguments?(grid_reference, postcode)
+
+      # Preference is to take it from the grid reference if available as it
+      # doesn't require a call to an external service, and is likely to be more
+      # accurate. If that can't be done or fails, then try the postcode
+      easting_and_northing_from_postcode(postcode) unless easting_and_northing_from_grid_reference(grid_reference)
```
Comment (cintamani) [inline app/services/waste_exemptions_engine/determine_easting_and_northing_service.rb]: [NIT]
```suggestion
easting_and_northing_from_grid_reference(grid_reference) || easting_and_northing_from_postcode(postcode)
```

File: app/services/waste_exemptions_engine/determine_easting_and_northing_service.rb [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+      @result
+    end
+
+    private
+
+    def valid_arguments?(grid_reference, postcode)
+      return false unless valid_argument?(grid_reference) || valid_argument?(postcode)
```
Comment (cintamani) [inline app/services/waste_exemptions_engine/determine_easting_and_northing_service.rb]:
```suggestion
      valid_argument?(grid_reference) || valid_argument?(postcode)
```

File: app/services/waste_exemptions_engine/determine_easting_and_northing_service.rb [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+      true
+    end
+
+    def valid_argument?(argument)
+      return false if argument.blank?
```
Comment (cintamani) [inline app/services/waste_exemptions_engine/determine_easting_and_northing_service.rb]:
```suggestion
      argument.present?
```

File: app/services/waste_exemptions_engine/determine_easting_and_northing_service.rb [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+      true
+    end
+
+    def easting_and_northing_from_grid_reference(grid_reference)
+      return false if grid_reference.blank?
```
Comment (cintamani) [inline app/services/waste_exemptions_engine/determine_easting_and_northing_service.rb]:
```suggestion
      return false unless valid_argument?(grid_reference)
```

**Comment by Cruikshanks** [comment]
@cintamani 
one_more_time

**Pull request merged**